semanticdb/metacp/src/main/scala/scala/meta/internal/javacp/Javacp.scala

```diff
@@ -155,23 +156,45 @@ object Javacp {
         }
         methodTypeParameters.foreach(buf += _)
 
-        val params =
-          if (isConstructor && hasOuterClassReference &&
-            // Guard against an empty parameter list, which seems to only happen
-            // in the JDK for java/util/regex/Pattern.class
-            method.signature.params.nonEmpty) {
-            // Drop the constructor argument that holds the reference to the outer class.
-            method.signature.params.tail
-          } else {
-            method.signature.params
-          }
+        val collectNonMandatedParams
+            : Tuple2[JavaTypeSignature, Int] => Option[(JavaTypeSignature, Option[String])] = {
+          case (_, 0) if isConstructor && hasOuterClassReference =>
+            // For JDK <21, mandated params holding the reference to the outer class
+            // in constructors cannot be detected with flags if `-parameters` was not
+            // passed to `javac`, so use an heuristic to detect such constructors and
+            // assume the first param is that reference.
+            // Note that in JDK >=21, the heuristic is unreliable, as the compiler
+            // does not synthetize the reference is not needed, but still injects an
+            // implicit param in the constructor, which is identified below via
+            // ACC_MANDATED.
+            None
+          case (sigParam, i) =>
+            val nodeParamOpt =
+              Option(method.node.parameters).flatMap(xs => Option(xs.get(i)))
+
+            nodeParamOpt match {
+              case Some(nodeParam)
+                  if (isConstructor && !isEnum &&
+                    (nodeParam.access.hasFlag(o.ACC_MANDATED) ||
+                      nodeParam.access.hasFlag(o.ACC_SYNTHETIC))) =>
```

Comment:
For some reason, JDK21 flags the implicit parameter of **private** inner classes constructors as `ACC_SYNTHETIC` (as opposed to `ACC_MANDATED` for non-private ones).

```
  com.javacp.Outer$PrivateInner(com.javacp.Outer, java.lang.Integer...);
    descriptor: (Lcom/javacp/Outer;[Ljava/lang/Integer;)V
    flags: (0x0080) ACC_VARARGS
    Code:
      stack=2, locals=3, args_size=3
         0: aload_0
         1: aload_1
         2: putfield      #1                  // Field this$0:Lcom/javacp/Outer;
         5: aload_0
         6: invokespecial #7                  // Method java/lang/Object."<init>":()V
         9: return
      LineNumberTable:
        line 12: 0
    MethodParameters:
      Name                           Flags
      this$0                         final synthetic
      values
```

```
  com.javacp.Outer$Inner(com.javacp.Outer, java.lang.Integer...);
    descriptor: (Lcom/javacp/Outer;[Ljava/lang/Integer;)V
    flags: (0x0080) ACC_VARARGS
    Code:
      stack=2, locals=3, args_size=3
         0: aload_0
         1: aload_1
         2: putfield      #1                  // Field this$0:Lcom/javacp/Outer;
         5: aload_0
         6: invokespecial #7                  // Method java/lang/Object."<init>":()V
         9: return
      LineNumberTable:
        line 5: 0
    MethodParameters:
      Name                           Flags
      this$0                         final mandated
      values
```

Adding this line alone fixes the `PrivateInner` test, but introduces a regression for enums (the `$enum$name` is stripped), therefore the ugly `!isEnum` above. Using the `startsWith(this$)` trick would make the tests pass, but that's because we have `-parameters` passed to `javac`. If we don't, we are only left with flags to infer what should be stripped or not.

I was hoping to run `ExpectSuite` without `-parameters`, but it might complicate the build unnecessarily. Since `-parameters` is still not the default on recent JDKs, we might want to remove it to make sure we cover the majority of cases?

Reply:
> I was hoping to run ExpectSuite without -parameters, but it might complicate the build unnecessarily. 

I just ran the test suite without `-parameters` locally as a sanity check. As expected, `ExpectSuite` shows that param names are missing (just as on `main`), but I can confirm that there are no other issues (since this PR does not rely on parameter names to skip implicit ones).

```diff
-com/javacp/Coin#`<init>`(). => private ctor <init>(param0: Int)
-  param0 => com/javacp/Coin#`<init>`().(param0)
+com/javacp/Coin#`<init>`(). => private ctor <init>($enum$name: Int)
+  $enum$name => com/javacp/Coin#`<init>`().($enum$name)
   Int => scala/Int#
-com/javacp/Coin#`<init>`().(param0) => param param0: Int
+com/javacp/Coin#`<init>`().($enum$name) => param $enum$name: Int
   Int => scala/Int#
   Int => scala/Int#
-com/javacp/Coin#valueOf(). => static method valueOf(param0: String): Coin
-  param0 => com/javacp/Coin#valueOf().(param0)
+com/javacp/Coin#valueOf(). => static method valueOf(name: String): Coin
+  name => com/javacp/Coin#valueOf().(name)
   String => java/lang/String#
   Coin => com/javacp/Coin#
-com/javacp/Coin#valueOf().(param0) => param param0: String
+com/javacp/Coin#valueOf().(name) => param name: String
   String => java/lang/String#
 com/javacp/MetacJava#StaticInner#NonStatic#`<init>`(). => ctor <init>()
-com/javacp/MetacJava#StaticInner#NonStatic#method(). => method method(param0: NonStatic): Unit
-  param0 => com/javacp/MetacJava#StaticInner#NonStatic#method().(param0)
+com/javacp/MetacJava#StaticInner#NonStatic#method(). => method method(e: NonStatic): Unit
+  e => com/javacp/MetacJava#StaticInner#NonStatic#method().(e)
   NonStatic => com/javacp/MetacJava#StaticInner#NonStatic#
   Unit => scala/Unit#
-com/javacp/MetacJava#StaticInner#NonStatic#method().(param0) => param param0: NonStatic
+com/javacp/MetacJava#StaticInner#NonStatic#method().(e) => param e: NonStatic
   NonStatic => com/javacp/MetacJava#StaticInner#NonStatic#
 com/javacp/MetacJava#`<init>`(). => ctor <init>()
-com/javacp/MetacJava#overload(). => method overload(param0: A): Unit
-  param0 => com/javacp/MetacJava#overload().(param0)
+com/javacp/MetacJava#overload(). => method overload(a: A): Unit
+  a => com/javacp/MetacJava#overload().(a)
   A => com/javacp/MetacJava#Overload1#A#
   Unit => scala/Unit#
-com/javacp/MetacJava#overload().(param0) => param param0: A
+com/javacp/MetacJava#overload().(a) => param a: A
   A => com/javacp/MetacJava#Overload1#A#
-com/javacp/MetacJava#overload(+1). => method overload(param0: A): Unit
-  param0 => com/javacp/MetacJava#overload(+1).(param0)
+com/javacp/MetacJava#overload(+1). => method overload(a: A): Unit
+  a => com/javacp/MetacJava#overload(+1).(a)
   A => com/javacp/MetacJava#Overload2#A#
   Unit => scala/Unit#
-com/javacp/MetacJava#overload(+1).(param0) => param param0: A
+com/javacp/MetacJava#overload(+1).(a) => param a: A
   A => com/javacp/MetacJava#Overload2#A#
-com/javacp/MetacJava#overload(+2). => static method overload(param0: A): Unit
-  param0 => com/javacp/MetacJava#overload(+2).(param0)
+com/javacp/MetacJava#overload(+2). => static method overload(a: A): Unit
+  a => com/javacp/MetacJava#overload(+2).(a)
   A => com/javacp/MetacJava#Overload3#A#
   Unit => scala/Unit#
-com/javacp/MetacJava#overload(+2).(param0) => param param0: A
+com/javacp/MetacJava#overload(+2).(a) => param a: A
   A => com/javacp/MetacJava#Overload3#A#
   Object => java/lang/Object#
-com/javacp/Outer#Inner#InnerInner#`<init>`(). => private[javacp] ctor <init>(param0: Integer*)
+com/javacp/Outer#Inner#InnerInner#`<init>`(). => private[javacp] ctor <init>(values: Integer*)
   javacp => com/javacp/
-  param0 => com/javacp/Outer#Inner#InnerInner#`<init>`().(param0)
+  values => com/javacp/Outer#Inner#InnerInner#`<init>`().(values)
   Integer => java/lang/Integer#
-com/javacp/Outer#Inner#InnerInner#`<init>`().(param0) => param param0: Integer*
+com/javacp/Outer#Inner#InnerInner#`<init>`().(values) => param values: Integer*
   Integer => java/lang/Integer#
   Unit => scala/Unit#
-com/javacp/Outer#Inner#`<init>`(). => private[javacp] ctor <init>(param0: Integer*)
+com/javacp/Outer#Inner#`<init>`(). => private[javacp] ctor <init>(values: Integer*)
   javacp => com/javacp/
-  param0 => com/javacp/Outer#Inner#`<init>`().(param0)
+  values => com/javacp/Outer#Inner#`<init>`().(values)
   Integer => java/lang/Integer#
-com/javacp/Outer#Inner#`<init>`().(param0) => param param0: Integer*
+com/javacp/Outer#Inner#`<init>`().(values) => param values: Integer*
   Integer => java/lang/Integer#
   Object => java/lang/Object#
-com/javacp/Outer#InnerUnusedOuter#`<init>`(). => private[javacp] ctor <init>(param0: Integer*)
+com/javacp/Outer#InnerUnusedOuter#`<init>`(). => private[javacp] ctor <init>(values: Integer*)
   javacp => com/javacp/
-  param0 => com/javacp/Outer#InnerUnusedOuter#`<init>`().(param0)
+  values => com/javacp/Outer#InnerUnusedOuter#`<init>`().(values)
   Integer => java/lang/Integer#
-com/javacp/Outer#InnerUnusedOuter#`<init>`().(param0) => param param0: Integer*
+com/javacp/Outer#InnerUnusedOuter#`<init>`().(values) => param values: Integer*
   Integer => java/lang/Integer#
   Object => java/lang/Object#
-com/javacp/Outer#PrivateInnerUnusedOuter#`<init>`(). => private[javacp] ctor <init>(param0: Integer*)
+com/javacp/Outer#PrivateInnerUnusedOuter#`<init>`(). => private[javacp] ctor <init>(values: Integer*)
   javacp => com/javacp/
-  param0 => com/javacp/Outer#PrivateInnerUnusedOuter#`<init>`().(param0)
+  values => com/javacp/Outer#PrivateInnerUnusedOuter#`<init>`().(values)
   Integer => java/lang/Integer#
-com/javacp/Outer#PrivateInnerUnusedOuter#`<init>`().(param0) => param param0: Integer*
+com/javacp/Outer#PrivateInnerUnusedOuter#`<init>`().(values) => param values: Integer*
   Integer => java/lang/Integer#
   javacp => com/javacp/
-com/javacp/Test#InnerShadowTypeParam#move(). => method move(param0: A, param1: B): Unit
-  param0 => com/javacp/Test#InnerShadowTypeParam#move().(param0)
+com/javacp/Test#InnerShadowTypeParam#move(). => method move(a: A, b: B): Unit
+  a => com/javacp/Test#InnerShadowTypeParam#move().(a)
   A => com/javacp/Test#InnerShadowTypeParam#[A]
-  param1 => com/javacp/Test#InnerShadowTypeParam#move().(param1)
+  b => com/javacp/Test#InnerShadowTypeParam#move().(b)
   B => com/javacp/Test#[B]
   Unit => scala/Unit#
-com/javacp/Test#InnerShadowTypeParam#move().(param0) => param param0: A
+com/javacp/Test#InnerShadowTypeParam#move().(a) => param a: A
   A => com/javacp/Test#InnerShadowTypeParam#[A]
-com/javacp/Test#InnerShadowTypeParam#move().(param1) => param param1: B
+com/javacp/Test#InnerShadowTypeParam#move().(b) => param b: B
   B => com/javacp/Test#[B]
   Int => scala/Int#
-com/javacp/Test#compareTo(). => final method compareTo(param0: B): Int
-  param0 => com/javacp/Test#compareTo().(param0)
+com/javacp/Test#compareTo(). => final method compareTo(b: B): Int
+  b => com/javacp/Test#compareTo().(b)
   B => com/javacp/Test#[B]
   Int => scala/Int#
-com/javacp/Test#compareTo().(param0) => param param0: B
+com/javacp/Test#compareTo().(b) => param b: B
   B => com/javacp/Test#[B]
   A => com/javacp/Test#[A]
-com/javacp/Test#genericParams(). => private[javacp] method genericParams(param0: A, param1: B): Unit
+com/javacp/Test#genericParams(). => private[javacp] method genericParams(a: A, b: B): Unit
   javacp => com/javacp/
-  param0 => com/javacp/Test#genericParams().(param0)
+  a => com/javacp/Test#genericParams().(a)
   A => com/javacp/Test#[A]
-  param1 => com/javacp/Test#genericParams().(param1)
+  b => com/javacp/Test#genericParams().(b)
   B => com/javacp/Test#[B]
   Unit => scala/Unit#
-com/javacp/Test#genericParams().(param0) => param param0: A
+com/javacp/Test#genericParams().(a) => param a: A
   A => com/javacp/Test#[A]
-com/javacp/Test#genericParams().(param1) => param param1: B
+com/javacp/Test#genericParams().(b) => param b: B
   B => com/javacp/Test#[B]
   Z => com/javacp/Z#
-com/javacp/Test#methodTypeParams(). => private[javacp] method methodTypeParams[C <: Integer](param0: C): Unit
+com/javacp/Test#methodTypeParams(). => private[javacp] method methodTypeParams[C <: Integer](c: C): Unit
   javacp => com/javacp/
   Integer => java/lang/Integer#
-  param0 => com/javacp/Test#methodTypeParams().(param0)
+  c => com/javacp/Test#methodTypeParams().(c)
   Unit => scala/Unit#
-com/javacp/Test#methodTypeParams().(param0) => param param0: C
+com/javacp/Test#methodTypeParams().(c) => param c: C
   C => com/javacp/Test#methodTypeParams().[C]
   Integer => java/lang/Integer#
-com/javacp/Test#overload(). => method overload(param0: Logger): Unit
-  param0 => com/javacp/Test#overload().(param0)
+com/javacp/Test#overload(). => method overload(a: Logger): Unit
+  a => com/javacp/Test#overload().(a)
   Logger => java/util/logging/Logger#
   Unit => scala/Unit#
-com/javacp/Test#overload().(param0) => param param0: Logger
+com/javacp/Test#overload().(a) => param a: Logger
   Logger => java/util/logging/Logger#
-com/javacp/Test#overload(+1). => method overload(param0: Logger): Unit
-  param0 => com/javacp/Test#overload(+1).(param0)
+com/javacp/Test#overload(+1). => method overload(a: Logger): Unit
+  a => com/javacp/Test#overload(+1).(a)
   Logger => com/javacp/Logger#
   Unit => scala/Unit#
-com/javacp/Test#overload(+1).(param0) => param param0: Logger
+com/javacp/Test#overload(+1).(a) => param a: Logger
   Logger => com/javacp/Logger#
   Unit => scala/Unit#
-com/javacp/Test#primitiveParams(). => private[javacp] method primitiveParams(param0: Int, param1: Long, param2: Float, param3: Double, param4: Short, param5: Byte, param6: Boolean, param7: Char): Unit
+com/javacp/Test#primitiveParams(). => private[javacp] method primitiveParams(a: Int, b: Long, c: Float, d: Double, e: Short, f: Byte, g: Boolean, h: Char): Unit
   javacp => com/javacp/
-  param0 => com/javacp/Test#primitiveParams().(param0)
+  a => com/javacp/Test#primitiveParams().(a)
   Int => scala/Int#
-  param1 => com/javacp/Test#primitiveParams().(param1)
+  b => com/javacp/Test#primitiveParams().(b)
   Long => scala/Long#
-  param2 => com/javacp/Test#primitiveParams().(param2)
+  c => com/javacp/Test#primitiveParams().(c)
   Float => scala/Float#
-  param3 => com/javacp/Test#primitiveParams().(param3)
+  d => com/javacp/Test#primitiveParams().(d)
   Double => scala/Double#
-  param4 => com/javacp/Test#primitiveParams().(param4)
+  e => com/javacp/Test#primitiveParams().(e)
   Short => scala/Short#
-  param5 => com/javacp/Test#primitiveParams().(param5)
+  f => com/javacp/Test#primitiveParams().(f)
   Byte => scala/Byte#
-  param6 => com/javacp/Test#primitiveParams().(param6)
+  g => com/javacp/Test#primitiveParams().(g)
   Boolean => scala/Boolean#
-  param7 => com/javacp/Test#primitiveParams().(param7)
+  h => com/javacp/Test#primitiveParams().(h)
   Char => scala/Char#
   Unit => scala/Unit#
-com/javacp/Test#primitiveParams().(param0) => param param0: Int
+com/javacp/Test#primitiveParams().(a) => param a: Int
   Int => scala/Int#
-com/javacp/Test#primitiveParams().(param1) => param param1: Long
+com/javacp/Test#primitiveParams().(b) => param b: Long
   Long => scala/Long#
-com/javacp/Test#primitiveParams().(param2) => param param2: Float
+com/javacp/Test#primitiveParams().(c) => param c: Float
   Float => scala/Float#
-com/javacp/Test#primitiveParams().(param3) => param param3: Double
+com/javacp/Test#primitiveParams().(d) => param d: Double
   Double => scala/Double#
-com/javacp/Test#primitiveParams().(param4) => param param4: Short
+com/javacp/Test#primitiveParams().(e) => param e: Short
   Short => scala/Short#
-com/javacp/Test#primitiveParams().(param5) => param param5: Byte
+com/javacp/Test#primitiveParams().(f) => param f: Byte
   Byte => scala/Byte#
-com/javacp/Test#primitiveParams().(param6) => param param6: Boolean
+com/javacp/Test#primitiveParams().(g) => param g: Boolean
   Boolean => scala/Boolean#
-com/javacp/Test#primitiveParams().(param7) => param param7: Char
+com/javacp/Test#primitiveParams().(h) => param h: Char
   Char => scala/Char#
   Unit => scala/Unit#
-com/javacp/Test#typeParams(). => private[javacp] method typeParams(param0: ArrayList[HashMap[A, Array[String]]], param1: Hashtable[String, B]): Unit
+com/javacp/Test#typeParams(). => private[javacp] method typeParams(a: ArrayList[HashMap[A, Array[String]]], b: Hashtable[String, B]): Unit
   javacp => com/javacp/
-  param0 => com/javacp/Test#typeParams().(param0)
+  a => com/javacp/Test#typeParams().(a)
   ArrayList => java/util/ArrayList#
   String => java/lang/String#
-  param1 => com/javacp/Test#typeParams().(param1)
+  b => com/javacp/Test#typeParams().(b)
   Hashtable => java/util/Hashtable#
   Unit => scala/Unit#
-com/javacp/Test#typeParams().(param0) => param param0: ArrayList[HashMap[A, Array[String]]]
+com/javacp/Test#typeParams().(a) => param a: ArrayList[HashMap[A, Array[String]]]
   ArrayList => java/util/ArrayList#
   String => java/lang/String#
-com/javacp/Test#typeParams().(param1) => param param1: Hashtable[String, B]
+com/javacp/Test#typeParams().(b) => param b: Hashtable[String, B]
   Hashtable => java/util/Hashtable#
   B => com/javacp/Test#[B]
-com/javacp/Test#vararg(). => private[javacp] method vararg(param0: Int, param1: String*): Unit
+com/javacp/Test#vararg(). => private[javacp] method vararg(a: Int, args: String*): Unit
   javacp => com/javacp/
-  param0 => com/javacp/Test#vararg().(param0)
+  a => com/javacp/Test#vararg().(a)
   Int => scala/Int#
-  param1 => com/javacp/Test#vararg().(param1)
+  args => com/javacp/Test#vararg().(args)
   String => java/lang/String#
   Unit => scala/Unit#
-com/javacp/Test#vararg().(param0) => param param0: Int
+com/javacp/Test#vararg().(a) => param a: Int
   Int => scala/Int#
-com/javacp/Test#vararg().(param1) => param param1: String*
+com/javacp/Test#vararg().(args) => param args: String*
   String => java/lang/String#
-com/javacp/Test#wildcard(). => method wildcard(param0: ArrayList[local_wildcard]): Unit
-  param0 => com/javacp/Test#wildcard().(param0)
+com/javacp/Test#wildcard(). => method wildcard(wildcard: ArrayList[local_wildcard]): Unit
+  wildcard => com/javacp/Test#wildcard().(wildcard)
   ArrayList => java/util/ArrayList#
   Unit => scala/Unit#
-com/javacp/Test#wildcard().(param0) => param param0: ArrayList[local_wildcard]
+com/javacp/Test#wildcard().(wildcard) => param wildcard: ArrayList[local_wildcard]
   ArrayList => java/util/ArrayList#
   local_wildcard => local_wildcard
-com/javacp/Test#wildcard(+1). => method wildcard(param0: Collection[local_wildcard], param1: Collection[local_wildcard]): Unit
-  param0 => com/javacp/Test#wildcard(+1).(param0)
+com/javacp/Test#wildcard(+1). => method wildcard(a: Collection[local_wildcard], b: Collection[local_wildcard]): Unit
+  a => com/javacp/Test#wildcard(+1).(a)
   Collection => java/util/Collection#
   local_wildcard => local_wildcard
-  param1 => com/javacp/Test#wildcard(+1).(param1)
+  b => com/javacp/Test#wildcard(+1).(b)
   Unit => scala/Unit#
-com/javacp/Test#wildcard(+1).(param0) => param param0: Collection[local_wildcard]
+com/javacp/Test#wildcard(+1).(a) => param a: Collection[local_wildcard]
   Collection => java/util/Collection#
   local_wildcard => local_wildcard
-com/javacp/Test#wildcard(+1).(param1) => param param1: Collection[local_wildcard]
+com/javacp/Test#wildcard(+1).(b) => param b: Collection[local_wildcard]
```